File: python/carts.py
import datetime
import sdk.python.entities
from sdk.python.entities import Property
from sdk.python.users import User
from sdk.python.domains import Domain
from sdk.python.addresses import Address
from sdk.python.invoices import Invoices
from sdk.python.shipments import Shipment
from sdk.python.companies import Company
from sdk.python.cart_items import CartItem


class Cart(sdk.python.entities.Entity):

    resource = '/carts/'
    json_name = 'cart'

    creation_date = Property(datetime.datetime)
    id = Property(int)
    ip = Property(str)
    token = Property(str)
    receiver_notes = Property(str)
    currency = Property(str)
    cost = Property(float)
    client = Property(User)
    client_company = Property(Company)
    domain = Property(Domain)
    shipment = Property(Shipment)
    invoice = Property(Invoices)
    cart_items = Property(CartItem)
    receiver_address = Property(Address)
    cart_items_subtotal_cost = Property(float)
    cart_items_tax_amount = Property(float)
    cart_items_total_cost = Property(float)
    subtotal_cost = Property(float)
    tax_amount = Property(float)
    total_cost = Property(float)


class Carts(sdk.python.entities.Resource):

    entity_class = Cart
    json_name = 'carts'


carts = Carts()

File: php/src/carts.php
<?php
declare(strict_types=1);

require_once 'entity.php';

class Cart extends Entity
{
    public static $json_name = 'cart';
    public static $resource = '/carts/';


    public function __construct()
    {
        $this->json_property('id', 'integer');
        $this->json_property('archived', 'integer');
        $this->json_property('creation_date', 'integer');
        $this->json_property('ip', 'string');
        $this->json_property('token', 'string');
        $this->json_property('receiver_notes', 'string');
        $this->json_property('currency', 'string');
        $this->json_property('cart_items_subtotal_cost', 'integer');
        $this->json_property('cart_items_tax_amount', 'integer');
        $this->json_property('cart_items_total_cost', 'integer');
        $this->json_property('shipment_total_cost', 'integer');
        $this->json_property('subtotal_cost', 'integer');
        $this->json_property('tax_amount', 'integer');
        $this->json_property('total_cost', 'integer');

        $this->json_property('client', 'User', $many = False,
            $recursive = True);
        $this->json_property('client_company', 'Company', $many = False,
            $recursive = True);
        $this->json_property('domain', 'Domain', $many = False,
            $recursive = True);
        $this->json_property('invoice', 'Invoice', $many = False,
            $recursive = True);
        $this->json_property('receiver_address', 'Address', $many = False,
        $recursive = True);
        $this->json_property('cart_items', 'CartItem', $many = True,
            $recursive = True);
        $this->json_property('shipment_groups', 'CartShipmentGroup', $many = True,
            $recursive = True);
    }

    public function requires_shipment()
    {
        if ($this->cart_items) {
            foreach($this->cart_items as $cart_item){
                if ($cart_item.requires_shipment()) {
                    return True;
                }
            }
            return False;
        } else {
            throw new Exception('cart_items is undefined, did you forget to embed it?');
        }
    }
}
